### ISSUE TITLE
Requiring vega-lite within a browserify package

### ISSUE THREAD

**Issue by arvind** [vega/vega-lite#881]
I have a `require('vega-lite')` line in one of my JavaScript files that gets bundled up with browserify. I don't have any shims because I would like to package all dependencies as part of a single built file. Unfortunately, this is throwing the following error during the browserify build process in my project:

```
Error: Cannot find module '../util' from '/project/node_modules/vega-lite'
    at /project/node_modules/browserify/node_modules/resolve/lib/async.js:55:21
    at load (/project/node_modules/browserify/node_modules/resolve/lib/async.js:69:43)
    at onex (/project/node_modules/browserify/node_modules/resolve/lib/async.js:92:31)
    at /project/node_modules/browserify/node_modules/resolve/lib/async.js:22:47
    at FSReqWrap.oncomplete (fs.js:95:15)
```

If I use a watchify process, the list of errors is even longer:

```
Error: Cannot find module '../util' from '/project/node_modules/vega-lite'
Error: Cannot find module '../time' from '/project/node_modules/vega-lite'
Error: Cannot find module './util' from '/project/node_modules/vega-lite'
Error: Cannot find module './time' from '/project/node_modules/vega-lite'
Error: Cannot find module './channel' from '/project/node_modules/vega-lite'
Error: Cannot find module '../channel' from '/project/node_modules/vega-lite'
Error: Cannot find module '../data' from '/project/node_modules/vega-lite'
Error: Cannot find module '../fielddef' from '/project/node_modules/vega-lite'
Error: Cannot find module '../encoding' from '/project/node_modules/vega-lite'
Error: Cannot find module './layout' from '/project/node_modules/vega-lite'
Error: Cannot find module '../mark' from '/project/node_modules/vega-lite'
Error: Cannot find module '../schema/schema' from '/project/node_modules/vega-lite'
Error: Cannot find module '../schema/schemautil' from '/project/node_modules/vega-lite'
Error: Cannot find module '../type' from '/project/node_modules/vega-lite'
Error: Cannot find module '../util' from '/project/node_modules/vega-lite'
Error: Cannot find module './time' from '/project/node_modules/vega-lite'
Error: Cannot find module '../util' from '/project/node_modules/vega-lite'
Error: Cannot find module '../type' from '/project/node_modules/vega-lite'
Error: Cannot find module '../channel' from '/project/node_modules/vega-lite'
Error: Cannot find module './time' from '/project/node_modules/vega-lite'
Error: Cannot find module './Model' from '/project/node_modules/vega-lite'
Error: Cannot find module './axis' from '/project/node_modules/vega-lite'
Error: Cannot find module './data' from '/project/node_modules/vega-lite'
Error: Cannot find module './facet' from '/project/node_modules/vega-lite'
Error: Cannot find module './legend' from '/project/node_modules/vega-lite'
Error: Cannot find module './marks' from '/project/node_modules/vega-lite'
Error: Cannot find module './scale' from '/project/node_modules/vega-lite'
Error: Cannot find module '../util' from '/project/node_modules/vega-lite'
Error: Cannot find module '../data' from '/project/node_modules/vega-lite'
Error: Cannot find module '../channel' from '/project/node_modules/vega-lite'
Error: Cannot find module './Model' from '/project/node_modules/vega-lite'
Error: Cannot find module '../fielddef' from '/project/node_modules/vega-lite'
Error: Cannot find module '../util' from '/project/node_modules/vega-lite'
Error: Cannot find module '../bin' from '/project/node_modules/vega-lite'
Error: Cannot find module '../channel' from '/project/node_modules/vega-lite'
Error: Cannot find module '../data' from '/project/node_modules/vega-lite'
Error: Cannot find module './time' from '/project/node_modules/vega-lite'
Error: Cannot find module '../type' from '/project/node_modules/vega-lite'
Error: Cannot find module '../util' from '/project/node_modules/vega-lite'
Error: Cannot find module '../channel' from '/project/node_modules/vega-lite'
Error: Cannot find module './axis' from '/project/node_modules/vega-lite'
Error: Cannot find module './scale' from '/project/node_modules/vega-lite'
Error: Cannot find module '../channel' from '/project/node_modules/vega-lite'
Error: Cannot find module '../mark' from '/project/node_modules/vega-lite'
Error: Cannot find module '../data' from '/project/node_modules/vega-lite'
Error: Cannot find module '../util' from '/project/node_modules/vega-lite'
Error: Cannot find module '../channel' from '/project/node_modules/vega-lite'
Error: Cannot find module './time' from '/project/node_modules/vega-lite'
Error: Cannot find module '../type' from '/project/node_modules/vega-lite'
Error: Cannot find module '../mark' from '/project/node_modules/vega-lite'
Error: Cannot find module '../channel' from '/project/node_modules/vega-lite'
Error: Cannot find module '../mark' from '/project/node_modules/vega-lite'
Error: Cannot find module '../type' from '/project/node_modules/vega-lite'
Error: Cannot find module './stack' from '/project/node_modules/vega-lite'
Error: Cannot find module '../util' from '/project/node_modules/vega-lite'
Error: Cannot find module '../aggregate' from '/project/node_modules/vega-lite'
Error: Cannot find module '../channel' from '/project/node_modules/vega-lite'
Error: Cannot find module '../data' from '/project/node_modules/vega-lite'
Error: Cannot find module './time' from '/project/node_modules/vega-lite'
Error: Cannot find module '../type' from '/project/node_modules/vega-lite'
Error: Cannot find module '../mark' from '/project/node_modules/vega-lite'
Error: Cannot find module '../util' from '/project/node_modules/vega-lite'
Error: Cannot find module '../util' from '/project/node_modules/vega-lite'
Error: Cannot find module '../channel' from '/project/node_modules/vega-lite'
Error: Cannot find module './type' from '/project/node_modules/vega-lite'
Error: Cannot find module './channel' from '/project/node_modules/vega-lite'
Error: Cannot find module './util' from '/project/node_modules/vega-lite'
Error: Cannot find module './compiler/time' from '/project/node_modules/vega-lite'
Error: Cannot find module './type' from '/project/node_modules/vega-lite'
Error: Cannot find module '../type' from '/project/node_modules/vega-lite'
Error: Cannot find module '../util' from '/project/node_modules/vega-lite'
Error: Cannot find module './config.stack.schema' from '/project/node_modules/vega-lite'
Error: Cannot find module './config.cell.schema' from '/project/node_modules/vega-lite'
Error: Cannot find module './config.marks.schema' from '/project/node_modules/vega-lite'
Error: Cannot find module './schemautil' from '/project/node_modules/vega-lite'
Error: Cannot find module '../util' from '/project/node_modules/vega-lite'
Error: Cannot find module './axis.schema' from '/project/node_modules/vega-lite'
Error: Cannot find module './legend.schema' from '/project/node_modules/vega-lite'
Error: Cannot find module './sort.schema' from '/project/node_modules/vega-lite'
Error: Cannot find module './fielddef.schema' from '/project/node_modules/vega-lite'
Error: Cannot find module './bin.schema' from '/project/node_modules/vega-lite'
Error: Cannot find module './scale.schema' from '/project/node_modules/vega-lite'
Error: Cannot find module '../aggregate' from '/project/node_modules/vega-lite'
Error: Cannot find module '../util' from '/project/node_modules/vega-lite'
Error: Cannot find module './schemautil' from '/project/node_modules/vega-lite'
Error: Cannot find module '../timeunit' from '/project/node_modules/vega-lite'
Error: Cannot find module '../type' from '/project/node_modules/vega-lite'
Error: Cannot find module '../util' from '/project/node_modules/vega-lite'
Error: Cannot find module './schemautil' from '/project/node_modules/vega-lite'
Error: Cannot find module '../type' from '/project/node_modules/vega-lite'
Error: Cannot find module './schemautil' from '/project/node_modules/vega-lite'
Error: Cannot find module './mark.schema' from '/project/node_modules/vega-lite'
Error: Cannot find module './config.schema' from '/project/node_modules/vega-lite'
Error: Cannot find module './data.schema' from '/project/node_modules/vega-lite'
Error: Cannot find module './encoding.schema' from '/project/node_modules/vega-lite'
Error: Cannot find module './fielddef.schema' from '/project/node_modules/vega-lite'
Error: Cannot find module '../util' from '/project/node_modules/vega-lite'
Error: Cannot find module '../aggregate' from '/project/node_modules/vega-lite'
Error: Cannot find module '../type' from '/project/node_modules/vega-lite'
Error: Cannot find module '../util' from '/project/node_modules/vega-lite'
Error: Cannot find module './aggregate' from '/project/node_modules/vega-lite'
Error: Cannot find module './timeunit' from '/project/node_modules/vega-lite'
Error: Cannot find module './type' from '/project/node_modules/vega-lite'
Error: Cannot find module './encoding' from '/project/node_modules/vega-lite'
Error: Cannot find module './mark' from '/project/node_modules/vega-lite'
Error: Cannot find module './compiler/Model' from '/project/node_modules/vega-lite'
Error: Cannot find module './channel' from '/project/node_modules/vega-lite'
Error: Cannot find module './encoding' from '/project/node_modules/vega-lite'
Error: Cannot find module './mark' from '/project/node_modules/vega-lite'
Error: Cannot find module './util' from '/project/node_modules/vega-lite'
Error: Cannot find module './util' from '/project/node_modules/vega-lite'
Error: Cannot find module './mark' from '/project/node_modules/vega-lite'
Error: Cannot find module './bin' from '/project/node_modules/vega-lite'
Error: Cannot find module './channel' from '/project/node_modules/vega-lite'
Error: Cannot find module './data' from '/project/node_modules/vega-lite'
Error: Cannot find module './encoding' from '/project/node_modules/vega-lite'
Error: Cannot find module './fielddef' from '/project/node_modules/vega-lite'
Error: Cannot find module './compiler/compiler' from '/project/node_modules/vega-lite'
Error: Cannot find module './schema/schema' from '/project/node_modules/vega-lite'
Error: Cannot find module './shorthand' from '/project/node_modules/vega-lite'
Error: Cannot find module './spec' from '/project/node_modules/vega-lite'
Error: Cannot find module './timeunit' from '/project/node_modules/vega-lite'
Error: Cannot find module './type' from '/project/node_modules/vega-lite'
Error: Cannot find module './validate' from '/project/node_modules/vega-lite'
Error: Cannot find module './util' from '/project/node_modules/vega-lite'
Error: Cannot find module 'd3-time' from '/project/node_modules/vega-lite'
[Error: write after end]
```

Investigating, it looks like the `main` property of `package.json` is specified as the built `vega-lite.js` file. This strikes me as a little odd for another reason: dependencies are packaged up as part of vega-lite, and I think requiring the built file prevents browserify from de-duping dependencies like it normally would do. Unsure how to proceed since the actual main file is in TypeScript and might require compiling?

**Comment by kanitw** [comment]
Interesting question, does this problem occur with linked version or latest npm release? 

If using from npm release, this should not be a problem?
For local linked version, I guess requiring building should be fine.  

Re: packaged dependencies, we only require data lib’s `util.js` and `generate.ts`.  
Given recent changes in datalib, https://github.com/vega/datalib/commit/494d764ac46acb41a4f517835d73f66ae9741b76.  

This would no longer require `d3-time` and `buffer` and thus make the duplicated dependency footprint relatively small (11kb).

**Comment by arvind** [comment]
I'm using the latest version of Vega-Lite from npm. You can reproduce this
issue with a simple browserify project, set up like Vega's for example but
without any shims, and require Vega-Lite within one of the files.
Several users have reported using Vega within their custom browserify
builds, I imagine the same will not be uncommon once Vega-Lite hits 1.0.
Using Vega within a custom browserify process is non-trivial, so there's a
wiki page on how to get it set up. Perhaps something similar is necessary
for Vega-Lite? Unfortunately, I'm not sufficiently familiar with the
TypeScript side of things to debug this further with Vega-Lite.

On Friday, December 18, 2015, Kanit Wongsuphasawat <notifications@github.com>
wrote:

> Interesting question, does this problem occur with linked version or
> latest npm release?
>
> If using from npm release, this should not be a problem?
> For local linked version, I guess requiring building should be fine.
>
> Re: packaged dependencies, we only require data lib’s util.js and
> generate.ts.
>
> Given recent changes in datalib, vega/datalib@494d764
> <https://github.com/vega/datalib/commit/494d764ac46acb41a4f517835d73f66ae9741b76>.
>
>
> This would no longer require d3-time and buffer and thus make the
> duplicated dependency footprint relatively small (11kb).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/vega/vega-lite/issues/881#issuecomment-165923656>.
>


-- 
-Arvind
-----
Arvind Satyanarayan <http://arvindsatya.com/>

**Comment by domoritz** [comment]
I don't rally understand browserify but I found that browserifying a browserified project has problems (see https://github.com/substack/node-browserify/issues/1310). https://stackoverflow.com/questions/28078425/browserifying-libraries-that-were-themselves-browserified-relative-paths-error/28176927#28176927 suggests as one solution derequire.

I got it working with

```js
var derequire = require('derequire');
var vl = derequire('vega-lite');
```

Another option is to require the root vl file

```js
var vl = require('vega-lite/src/vl.js');
```

I'm closing this now. @arvind let me know whether this works for you. Also, if you think that we should change something about our set up to prevent this problem, let me know.
